frontends/common/model.h

```diff
@@ -62,4 +65,103 @@ class Model {
 
 }  // namespace Model
 
+namespace P4 {
+
+using ::Model::Elem;
+using ::Model::Type_Model;
+using ::Model::Param_Model;
+
+// Block has a name and a collection of elements
+template<typename T>
+struct Block_Model : public Type_Model {
+    std::vector<T> elems;
+    explicit Block_Model(cstring name) :
+        ::Model::Type_Model(name) {}
+};
+
+/// Enum_Model : Block_Model<Elem> : Type_Model
+struct Enum_Model : public Block_Model<Elem> {
+    ::Model::Type_Model type;
+    explicit Enum_Model(cstring name) :
+        Block_Model(name), type("Enum") {}
+};
+
+/// Parser_Model : Block_Model<Param_Model> : Type_Model
+struct Parser_Model : public Block_Model<Param_Model> {
+    ::Model::Type_Model type;
+    explicit Parser_Model(cstring name) :
+        Block_Model<Param_Model>(name), type("Parser") {}
+};
+
+/// Control_Model : Block_Model<Param_Model> : Type_Model
+struct Control_Model : public Block_Model<Param_Model> {
+    ::Model::Type_Model type;
+    explicit Control_Model(cstring name) :
+        Block_Model<Param_Model>(name), type("Control") {}
+};
+
+/// Method_Model : Block_Model<Param_Model> : Type_Model
+struct Method_Model : public Block_Model<Param_Model> {
+    ::Model::Type_Model type;
+    explicit Method_Model(cstring name) :
+        Block_Model<Param_Model>(name), type("Method") {}
+};
+
+/// Extern_Model : Block_Model<Method_Model> : Type_Model
+struct Extern_Model : public Block_Model<Method_Model> {
```

Comment:
How do you distinguish between decls and methods in an extern? Can externs have only Methods?